### PR TITLE
feat: oracle reader access control

### PR DESCRIPTION
**Pull request by xenide**
# WIP 

## Solution

- currently we have the following roles. Do you think it makes sense to refactor and use an auth library? 
1. `factory`
2. `assetManager`
3. `authorizedOracleCaller`

- in this architecture there will be no "custom" oracle callers, only the one that is specified by the factory

### CODE REVIEW

File: src/curve/stable/StablePair.sol [hunk 1]
```diff
@@ -11,7 +11,7 @@ import { GenericFactory } from "src/GenericFactory.sol";
 import { IReservoirCallee } from "src/interfaces/IReservoirCallee.sol";
 import { StableMath } from "src/libraries/StableMath.sol";
 import { StableOracleMath } from "src/libraries/StableOracleMath.sol";
-import { ReservoirPair } from "src/ReservoirPair.sol";
+import { ReservoirPair, Observation } from "src/ReservoirPair.sol";
```
Comment (OliverNChalk) [inline src/curve/stable/StablePair.sol:14]: Interesting, is it necessary to re-declare the Observation import all the time?

File: src/oracle/OracleWriter.sol [hunk 2]
```diff
@@ -17,19 +17,34 @@ abstract contract OracleWriter is Pair, IOracleWriter {
     uint256 internal constant MAX_CHANGE_PER_SEC = 0.01e18;
     string internal constant ALLOWED_CHANGE_NAME = "Shared::allowedChangePerSecond";
 
-    Observation[65536] public observations;
+    Observation[65536] internal _observations;
     uint16 public index = type(uint16).max;
 
+    mapping(address => bool) authorizedCallers;
```
Comment (OliverNChalk) [inline src/oracle/OracleWriter.sol]: I don't like using a mapping, can we just have a single address called `OracleReader` that is authorized? That address can then be changed to support contract upgrading.

In general we should model our system like MakerDAO does, small components that can be swapped out by simply changing an address.

File: src/oracle/OracleWriter.sol [hunk 3]
```diff
@@ -17,19 +17,34 @@ abstract contract OracleWriter is Pair, IOracleWriter {
...
         setAllowedChangePerSecond(lAllowedChangePerSecond);
     }
 
+    function observations(uint256 aIndex) external view authorized returns(Observation memory rObservation) {
+        return _observations[aIndex];
+    }
+
+    function setAuthorized(address aCaller, bool aAuthorized) external onlyFactory {
+        authorizedCallers[aCaller] = aAuthorized;
+    }
```
Comment (OliverNChalk) [inline src/oracle/OracleWriter.sol]: When you move array from a mapping model, you should then support pulling the new address from the factory, similar to fees

Comment (OliverNChalk) [reply src/oracle/OracleWriter.sol]: Also, does this design mean we need to configure every single pair to be authorized? Seems very intense

**Review by OliverNChalk — CHANGES_REQUESTED** [review]

File: src/curve/stable/StablePair.sol (hunk 1 above)
Comment (xenide) [reply src/curve/stable/StablePair.sol:14]: yes seems to be the case if we want to use the struct

File: src/oracle/OracleCaller.sol [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    constructor(address aOwner) Owned(aOwner) {} // solhint-disable-line no-empty-blocks
+
+    function observation(IOracleWriter aPair, uint256 aIndex) external view returns (Observation memory rObservation) {
+        require(whitelist[msg.sender], "OC: NOT_WHITELISTED");
+        rObservation = aPair.observation(aIndex);
+    }
+
+    function whitelistAddress(address aCaller, bool aWhitelist) external onlyOwner {
+        if (whitelist[aCaller] != aWhitelist) {
```
Comment (OliverNChalk) [inline src/oracle/OracleCaller.sol]: I would remove this check, increases gas cost for smart humans and only save gas for silly humans (with no loss of funds possible)

File: src/oracle/OracleWriter.sol [hunk 5]
```diff
@@ -16,18 +16,38 @@ abstract contract OracleWriter is Pair, IOracleWriter {
     // 100 basis points per second which is 60% per minute
     uint256 internal constant MAX_CHANGE_PER_SEC = 0.01e18;
     string internal constant ALLOWED_CHANGE_NAME = "Shared::allowedChangePerSecond";
+    string internal constant ORACLE_CALLER_NAME = "Shared::oracleCaller";
 
-    Observation[65536] public observations;
+    Observation[65536] public _observations;
```
Comment (OliverNChalk) [inline src/oracle/OracleWriter.sol]: Should be private right?
```suggestion
    Observation[65536] private _observations;
```

File: src/oracle/OracleWriter.sol [hunk 6]
```diff
@@ -16,18 +16,38 @@ abstract contract OracleWriter is Pair, IOracleWriter {
...
+        rObservation = _observations[aIndex];
+    }
+
+    function updateOracleCaller() public {
+        address lNewCaller = factory.read(ORACLE_CALLER_NAME).toAddress();
+        if (lNewCaller != oracleCaller) {
+            emit OracleCallerChanged(oracleCaller, lNewCaller);
+            oracleCaller = lNewCaller;
+        }
     }
```
Comment (OliverNChalk) [inline src/oracle/OracleWriter.sol:47]: Here the check is fine to avoid spammers

File: src/oracle/OracleWriter.sol [hunk 7]
```diff
@@ -16,18 +16,38 @@ abstract contract OracleWriter is Pair, IOracleWriter {
...
     // to mitigate oracle manipulation attacks in the face of post-merge ETH
     uint256 public allowedChangePerSecond;
     uint256 public prevClampedPrice;
 
+    address public oracleCaller;
+
+    modifier onlyOracleCaller() {
+        require(msg.sender == oracleCaller, "OW: NOT_ORACLE_CALLER");
+        _;
+    }
```
Comment (OliverNChalk) [inline src/oracle/OracleWriter.sol]: I think its more readable & less code to just inline this check, agree? Vyper actually does not support modifiers btw

**Review by OliverNChalk — COMMENTED** [review]
Looking good, just some minor comments and what looks like 1 bug

File: src/oracle/OracleWriter.sol (hunk 5 above)
Comment (xenide) [reply src/oracle/OracleWriter.sol]: that's right. fixed

File: src/oracle/OracleCaller.sol (hunk 4 above)
Comment (xenide) [reply src/oracle/OracleCaller.sol]: alright fixed

File: src/oracle/OracleWriter.sol (hunk 7 above)
Comment (xenide) [reply src/oracle/OracleWriter.sol]: yeah I thought about it too. It is indeed fewer lines of code to inline it (which saves us money). 

Btw how is vyper relevant to our code here? We ain't using vyper

Comment (OliverNChalk) [reply src/oracle/OracleWriter.sol]: One should study history and the approach of others when attempting to determine the best path for themselves

**Review by OliverNChalk — APPROVED** [review]

**Pull request merged**